src/main/java/at/ac/tuwien/kr/alpha/grounder/parser/ParseTreeVisitor.java

```diff
@@ -3,42 +3,92 @@
 import at.ac.tuwien.kr.alpha.antlr.ASPCore2BaseVisitor;
 import at.ac.tuwien.kr.alpha.antlr.ASPCore2Lexer;
 import at.ac.tuwien.kr.alpha.antlr.ASPCore2Parser;
-import at.ac.tuwien.kr.alpha.common.BasicPredicate;
-import at.ac.tuwien.kr.alpha.common.Program;
-import at.ac.tuwien.kr.alpha.common.Rule;
+import at.ac.tuwien.kr.alpha.common.*;
 import at.ac.tuwien.kr.alpha.common.atoms.Atom;
 import at.ac.tuwien.kr.alpha.common.atoms.BasicAtom;
-import at.ac.tuwien.kr.alpha.common.atoms.BuiltinAtom;
 import at.ac.tuwien.kr.alpha.common.atoms.Literal;
+import at.ac.tuwien.kr.alpha.common.predicates.BasicPredicate;
+import at.ac.tuwien.kr.alpha.common.predicates.Predicate;
+import at.ac.tuwien.kr.alpha.common.predicates.TotalOrder;
 import at.ac.tuwien.kr.alpha.common.terms.*;
 import org.antlr.v4.runtime.RuleContext;
 import org.antlr.v4.runtime.tree.TerminalNode;
 
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.LinkedList;
-import java.util.List;
+import java.util.*;
+
+import static java.util.Collections.emptyList;
 
 /**
  * Copyright (c) 2016, the Alpha Team.
  */
 public class ParseTreeVisitor extends ASPCore2BaseVisitor<Object> {
+	private final Map<String, Predicate> externals;
+	private final boolean acceptVariables;
 
 	private Program inputProgram;
 	private boolean isCurrentLiteralNegated;
 
+	public ParseTreeVisitor(Map<String, Predicate> externals) {
+		this(externals, true);
+	}
+
+	public ParseTreeVisitor(Map<String, Predicate> externals, boolean acceptVariables) {
+		this.externals = externals;
+		this.acceptVariables = acceptVariables;
+	}
+
 	private void notSupportedSyntax(RuleContext ctx) {
 		throw new UnsupportedOperationException("Unsupported syntax encountered: " + ctx.getText());
 	}
 
+	public Program translate(ASPCore2Parser.ProgramContext input) {
```

Comment:
What is translated here? ParseProgram might be a better name.

Reply:
Ahh, right, my understanding at this point was wrong.
Btw, your explanation would make a good javadoc for this method. :-)